README.md
NetSquid-BoardGames (0.0.1)
================================

Description
-----------

This is a user contributed _snippet_ for the [NetSquid quantum network simulator](https://netsquid.org).

For multiplayer board games with a Task Master based on quantum operations. THIS IS A WORK IN PROGRESS.
The board games are such that the users cannot look at the contents of the board at all times. The board remains with the TaskMaster. The players have to calculate the move that their opponent might have taken and plan their moves accordingly. The place to operate on(qubit) is given to a player by their opponent. 

Installation
------------

See the [INSTALL file](INSTALL.md) for instruction of how to install this snippet.

Documentation
-------------

To build and see the docs see the [docs README](docs/README.md).

Usage
-----

**TEMPLATE**: briefly describe how to use your package.

Contributors
------------

**TEMPLATE**: specify who contributed to your package and their contact details.

License
-------

**TEMPLATE**: specify the license applicable to your package.

The NetSquid-SnippetTemplate has the following license:

> Copyright 2020 QuTech (TUDelft and TNO)
>
>   Licensed under the Apache License, Version 2.0 (the "License");
>   you may not use this file except in compliance with the License.
>   You may obtain a copy of the License at
>
>     http://www.apache.org/licenses/LICENSE-2.0
>
>   Unless required by applicable law or agreed to in writing, software
>   distributed under the License is distributed on an "AS IS" BASIS,
>   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>   See the License for the specific language governing permissions and
>   limitations under the License.


CHANGELOG.md
CHANGELOG
=========

2020-08-24 (0.0.1)
------------------

Example changelog entry

- Created this snippet


tests/test_samplemodule.py
# This file uses NumPy style docstrings: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

"""Unit tests for the sample module.

This file will be discovered by ``python setup.py test``.

"""
import pytest
import numpy as np
from netsquid.qubits.ketstates import s0, s1, s00, s11, b00
from netsquid.qubits.qubitapi import fidelity
from netsquid_boardgames.samplemodule import create_ghz_qubits


@pytest.mark.parametrize("num_qubits, expected_state", [
    (2, b00),
    (3, (np.kron(s0, s00) + np.kron(s1, s11)) / np.sqrt(2)),
    (4, (np.kron(s00, s00) + np.kron(s11, s11)) / np.sqrt(2)),
])
def test_create_ghz_state(num_qubits, expected_state):
    """Test creation of GHZ states."""
    qubits = create_ghz_qubits(num_qubits)
    assert fidelity(qubits, expected_state) == pytest.approx(1.)


@pytest.mark.parametrize("num_qubits, expected_error", [
    (3., TypeError),
    ("3", TypeError),
    (0, ValueError),
    (1, ValueError),
])
def test_faulty(num_qubits, expected_error):
    with pytest.raises(expected_error):
        create_ghz_qubits(num_qubits)


examples/example1.py
import random

import netsquid as ns
from netsquid.nodes import Node
from netsquid.components import QuantumChannel, Channel
from netsquid.nodes import DirectConnection
from netsquid_boardgames.protocols import TaskMasterProtocol, PlayerProtocol

def main():
	ns.sim_reset()
	tm_node = Node(name="TaskMaster")
	p1_node = Node(name="PlayerOne")
	p0_node = Node(name="PlayerZero")

	
	cht0q = QuantumChannel(name="qchannel[tm to p0]")
	cht1q = QuantumChannel(name="qchannel[tm to p1]")
	ch0tq = QuantumChannel(name="qchannel[p0 to tm]")
	ch1tq = QuantumChannel(name="qchannel[p1 to tm]")

	ch0tc = Channel(name="channel[p0 to tm]")
	ch1tc = Channel(name="channel[p1 to tm]")

	tm_node.add_ports('cIO0')
	tm_node.add_ports('cIO1')
	p0_node.add_ports('cPtoTM')
	p1_node.add_ports('cPtoTM')

	ch0tc.ports['send'].connect(p0_node.ports['cPtoTM'])
	ch0tc.ports['recv'].connect(tm_node.ports['cIO0'])
	ch1tc.ports['send'].connect(p1_node.ports['cPtoTM'])
	ch1tc.ports['recv'].connect(tm_node.ports['cIO1'])

	
	qconnection0 = DirectConnection(name="qconn[tm|p0]", channel_AtoB=cht0q, channel_BtoA=ch0tq)
	qconnection1 = DirectConnection(name="qconn[tm|p1]", channel_AtoB=cht1q, channel_BtoA=ch1tq)

	tm_node.connect_to(remote_node=p0_node, connection=qconnection0, local_port_name="IOp0", remote_port_name="qubitIO")
	tm_node.connect_to(remote_node=p1_node, connection=qconnection1, local_port_name="IOp1", remote_port_name="qubitIO")


	tm_protocol = TaskMasterProtocol(tm_node, 25)
	p0_protocol = PlayerProtocol(p0_node, 0)
	p1_protocol = PlayerProtocol(p1_node, 1)

	p0_protocol.start()
	p1_protocol.start()
	tm_protocol.start()

	ns.sim_run()

if __name__ == "__main__":
     main()



netsquid_boardgames/protocols.py
import netsquid as ns
from netsquid.protocols import NodeProtocol
from netsquid.components import QuantumMemory

class TaskMasterProtocol(NodeProtocol):
	def __init__(self, node, memory_size):
		super().__init__(node)
		self.node = node
		self.memory_size = memory_size
		self.memory = QuantumMemory("TaskMasterMemory", num_positions=memory_size)
		mem = ns.qubits.create_qubits(memory_size)
		self.memory.put(mem)
		
		self.index = random.randint(0,24)

	def run(self):
		while True:
			print("current index is :", self.index)
			self.node.ports["IOp0"].tx_output(self.memory.pop(positions=self.index))

			yield self.await_port_input(self.node.ports["IOp0"])
			qbit_return = self.node.ports["IOp0"].rx_input()
			self.memory.put(qbit_return, positions=self.index)

			yield self.await_port_input(self.node.ports["cIO0"])
			move0 = self.node.ports["cIO0"].rx_input()
			self.index = int(move0)

			self.node.ports["IOp1"].tx_output(self.memory.pop(positions=self.index))

			yield self.await_port_input(self.node.ports["IOp1"])
			qbit_return = self.node.ports["IOp1"].rx_input()
			self.memory.put(qbit_return, positions=self.index)

			yield self.await_port_input(self.node.ports["cIO1"])
			move1 = self.node.ports["cIO1"].rx_input()
			self.index = int(move1)


class PlayerProtocol(NodeProtocol):
	def __init__(self, node, sign):
		super().__init__(node)
		self.node = node
		self.sign = sign
		print("You are player ", sign)

	def run(self):
		while True:
			yield self.await_port_input(self.node.ports["qubitIO"])
			message = self.node.ports["qubitIO"].rx_input()
			qubit = message.items[0]
			print(qubit)
			operation = None
			ops = ['X','Y','Z','H','K','S','T','I']
			#Ask the player what they want to do with the qubit
			while operation not in ops:
				operation = input("Enter the operation you want to perform on qubit: ")
				if operation not in ops:
					print("Invalid operation. Choose from ", ops)
			ns.qubits.operate(qubit, getattr(ns, operation))

			#send
			self.node.ports["cPtoTM"].tx_output(qubit)

			#ask the player the qubit to send to the opponent
			move = -1
			while not 0<= move <25:
				move = int(input("Enter the qubit index to send to opponent: "))
				if not 0<= move < 25:
					print("Invalid move. Choose a number from range [0,24]")
			
			#send the move
			self.node.ports["cPtoTM"].tx_output(move)
